inc/namespace.php

```diff
@@ -246,6 +262,130 @@ function noop_wp_query_found_rows_on_failed_ep_request( string $sql, WP_Query $q
+ *
+ * @param array $query The Elasticsearch query.
+ * @param string|null $type The remote request type.
+ * @return void
+ */
+function on_delete_index( $query, ?string $type ) {
+	if ( $type !== 'delete_index' ) {
+		return;
+	}
+	// Set the version if not already present.
+	if ( ! get_index_version() ) {
+		set_index_version( 1 );
+	}
+}
+
+/**
+ * Set the index version for the current site.
+ *
+ * @param integer $version The version number.
+ * @return void
+ */
+function set_index_version( int $version ) {
+	update_option( 'altis_search_index_version', $version );
+}
+
+/**
+ * Get the index version for the current site.
+ *
+ * @return int|null
+ */
+function get_index_version() : ?int {
+	return get_option( 'altis_search_index_version', null );
+}
+
+/**
+ * ElasticPress adds the indexable object type to index names. We can maintain backwards
+ * compatibility by filtering the posts indexable index name to remove this type.
+ *
+ * @param string $index The index name.
+ * @return string
+ */
+function filter_index_name( string $index ) : string {
+	// Back compat for Altis v3 & ElasticPress 2.x
+	// Version 3 of ElasticPress introduces Indexables allowing for user
+	// and term search integration. The new index names follow the pattern
+	// <site>-<indexable>-<blog-id> instead of <site>-<blog-id>.
+	$index_version = get_index_version();
+	if ( ! $index_version && strpos( $index, '-post' ) !== false ) {
+		$old_index = str_replace( '-post', '', $index );
+		if ( Elasticsearch::factory()->index_exists( $old_index ) ) {
+			return $old_index;
+		}
+	}
+
+	return $index;
+}
+
+/**
+ * The documents ingest pipeline does not need to be site specific
+ * as it is always the same.
+ *
+ * @return string
+ */
+function filter_documents_pipeline_id( string $id ) : string {
+	if ( ! get_index_version() ) {
+		return $id;
+	}
+	return 'attachments';
+}
+
+/**
+ * Ensure ElasticPress requests do not impact on indexes the plugin does not manage.
+ *
+ * @param string $url The full Elasticsearch request URL.
+ * @param integer $failures Number of failures.
+ * @param string $host Host name.
+ * @param string $path Request path.
+ * @param array $args Remote request arguments.
+ * @return string
+ */
+function protect_non_ep_indexes( string $url, int $failures, string $host, string $path, array $args ) : string {
+
+	// If the path uses a wildcard protect the following index patterns.
+	$protected_index_patterns = [
+		'.kibana',
+		'analytics*',
+	];
+
+	/**
+	 * Filter the protected index patterns. This should be an array of
+	 * Elasticsearch index patterns that ElasticPress is not allowed to
+	 * interact with.
+	 *
+	 * @param array $protected_index_patterns The patterns ElasticPress should ignore.
+	 * @param string $path The remote request path.
+	 * @param array $args The remote request args.
+	 */
+	$protected_index_patterns = apply_filters( 'altis.search.protected_index_patterns', $protected_index_patterns, $path, $args );
```

Comment:
Can we switch this approach to an include instead of an exclude?

Reply:
Hmm, maybe. That would be safer. It would be best if we set a known index prefix for all ElasticPress managed indexes. We could do that using the `EP_PREFIX` constant, or filter I suppose but again it'd require a reindex to work...

I'd want to avoid having to build up a list of all EP indexes by looping over all sites and all indexables to get the index names basically. Will have a think.